### PR TITLE
lib: os: Introduce support for CRC32C algorithm

### PR DESCRIPTION
**Pull request by rgundi**
This introduces the support for CRC32C (Castagnoli) algorithm.
The generator polynomial used is 0x1EDC6F41UL.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

### CODE REVIEW

File: lib/os/crc32c_sw.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2021 Workaround GmbH.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sys/crc.h>
+
+/* crc table generated from polynomial 0x1EDC6F41UL (Castagnoli) */
+static const uint32_t crc32c_table[256] = {
```
Comment (jukkar) [inline lib/os/crc32c_sw.c:10]: You need to implement this crc func without this large lookup table to save flash, just like how it is done for crc32.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: lib/os/crc32c_sw.c (hunk 1 above)
Comment (rgundi) [reply lib/os/crc32c_sw.c:10]: But, isn't that a hit on performance? For the application where i am using it, performance is critical. We have to calculate the CRC for tens of MBs of data. And we have lots of flash space left.

Comment (jukkar) [reply lib/os/crc32c_sw.c:10]: That is your use case, but there are other use cases too. Currently all the other crc functions are semi-optimized for smaller lookup tables or there is no lookup table at all. For example `crc32_ieee_update()` does have small lookup table of 64 bytes.

Comment (rgundi) [reply lib/os/crc32c_sw.c]: Made changes to use a smaller lookup of only 16 bytes.

**Comment by jukkar** [comment]
Please add unit test to `tests/unit/crc` for this.

**Comment by rgundi** [comment]
> Please add unit test to `tests/unit/crc` for this.

Done.

**Review by jukkar — APPROVED** [review]
LGTM

File: tests/unit/crc/main.c [hunk 2]
```diff
@@ -9,8 +9,19 @@
...
 
+void test_crc32c(void)
+{
+	uint8_t test1[] = { 'A' };
+	uint8_t test2[] = { '1', '2', '3', '4', '5', '6', '7', '8', '9' };
+	uint8_t test3[] = { 'Z', 'e', 'p', 'h', 'y', 'r' };
+
+	zassert_equal(crc32_c(0, test1, sizeof(test1), 1, 1), 0xE16DCDEE, NULL);
+	zassert_equal(crc32_c(0, test2, sizeof(test2), 1, 1), 0xE3069283, NULL);
+	zassert_equal(crc32_c(0, test3, sizeof(test3), 1, 1), 0xFCDEB58D, NULL);
```
Comment (albertofloyd) [inline tests/unit/crc/main.c:23]: Not sure if part of regular test cases, but would be great if there is a test case where first_pkt = 0 to increase coverage.

Comment (rgundi) [reply tests/unit/crc/main.c]: Thanks for pointing it out. It's a shame i missed it. Done now.

File: lib/os/crc32c_sw.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	if (first_pkt) {
+		crc = CRC32C_INIT;
+	}
+
+	for (size_t i = 0; i < len; i++) {
+		crc = crc32c_table[(crc ^ data[i]) & 0x0F] ^ (crc >> 4);
+		crc = crc32c_table[(crc ^ (data[i] >> 4)) & 0x0F] ^ (crc >> 4);
+	}
+
+	return last_pkt?(crc ^ CRC32C_XOR_OUT):crc;
```
Comment (carlescufi) [inline lib/os/crc32c_sw.c:39]:
```suggestion
	return last_pkt ? (crc ^ CRC32C_XOR_OUT) : crc;
```

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: lib/os/crc32c_sw.c (hunk 3 above)
Comment (rgundi) [reply lib/os/crc32c_sw.c]: Done

Comment (rgundi) [reply lib/os/crc32c_sw.c]: @carlescufi I would have expected the checkpatch to catch this though.

**Review by dcpleung — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**